### PR TITLE
(WIP) isom-1944 security writing tests for routes (auth + perms)

### PR DESCRIPTION
**Pull request by adriangohjw**
## Problem

Closes [insert issue #]

## Solution

**Breaking Changes**

- [ ] Yes - this PR contains breaking changes
  - Details ...
- [ ] No - this PR is backwards compatible

**Features**:

- Details ...

**Improvements**:

- Details ...

**Bug Fixes**:

- Details ...

## Before & After Screenshots

**BEFORE**:

**AFTER**:

## Tests

**New scripts**:

- `script` : script details

**New dependencies**:

- `dependency` : dependency details

**New dev dependencies**:

- `dependency` : dependency details

### CODE REVIEW

**Comment by adriangohjw** [comment]
bugbot run

**Review by seaerchin — COMMENTED** [review]
**NOTE: this is a first pass review from claude**

⏺ Code Review: PR #1371 - Security Testing for Routes (Auth + Perms)

  Overview

  This PR adds comprehensive authentication and authorization tests for route handlers across multiple modules (collection, folder, page, resource, user, and me). The changes include 1,744 additions and 291 deletions, focusing on security testing to ensure proper access controls.

  Code Quality & Style

  ✅ Strengths

  - Comprehensive Coverage: Tests cover authentication (401) and authorization (403) scenarios across all major modules
  - Consistent Test Structure: Well-organized test patterns with clear arrange/act/assert structure
  - Security Focus: Addresses critical security concerns by validating permissions at route level
  - Proper Error Handling: Tests verify correct TRPC error codes and messages

  ⚠ Areas for Improvement

  1. Code Duplication in Test Setup
  ```ts
  // Repeated across multiple test files
  it("should throw 401 if not logged in", async () => {
    const result = unauthedCaller.someMethod({...})
    await expect(result).rejects.toThrowError(
      new TRPCError({ code: "UNAUTHORIZED" })
    )
  })
```

  3. Missing Test Utilities
  Tests could benefit from shared utilities for common patterns like permission setup and error assertions.

  4. Schema Changes Need Validation
  - Removal of getPermissionsInputSchema needs verification that it's not breaking existing functionality
  - New schemas like getRolesForSchema should be validated against usage patterns

  Specific Suggestions

  1. Extract Common Test Utilities

```ts
  // tests/integration/helpers/security.ts
  export const expectUnauthorized = (promise: Promise<any>) =>
    expect(promise).rejects.toThrowError(
      new TRPCError({ code: "UNAUTHORIZED" })
    )

  export const expectForbidden = (promise: Promise<any>) =>
    expect(promise).rejects.toThrowError(
      new TRPCError({
        code: "FORBIDDEN",
        message: "You do not have sufficient permissions to perform this action"
      })
    )
```

  2. Improve Query Security

```ts
  // apps/studio/src/server/modules/folder/folder.router.ts:291-294
  const { title } = await db
    .selectFrom("Resource")
    .where("Resource.siteId", "=", siteId) // ✅ Good addition
    .where("Resource.id", "=", resourceId)
    .select("title")
    .executeTakeFirstOrThrow()
```

  3. Schema Migration Validation

  The removal of user.getPermissions route and replacement with resource.getRolesFor needs careful validation:
```ts
  // Before: UserManagementContext.tsx:25
  const { data: roles } = trpc.user.getPermissions.useQuery({ siteId })

  // After: UserManagementContext.tsx:25-28
  const { data: roles } = trpc.resource.getRolesFor.useQuery({
    siteId,
    resourceId: null,
  })
```

  Code Correctness

  - ✅ Proper authentication checks added to all routes
  - ✅ Authorization logic correctly implemented
  - ✅ Database queries properly scoped by siteId
  - ⚠ Schema migration needs validation for breaking changes

  Performance Implications

  - ✅ No significant performance concerns
  - ✅ Database queries are properly indexed fields (siteId, id)
  - ✅ Permission checks are efficient

  Test Coverage

  - ✅ Excellent coverage for auth/authz scenarios
  - ✅ Both positive and negative test cases
  - ✅ Edge cases covered (non-existent resources, wrong permissions)
  - ❌ Missing integration tests for schema migration

  Security Considerations

  - ✅ Major Security Improvement: Adds missing authorization checks
  - ✅ Proper separation of read/write/publish permissions
  - ✅ Prevents unauthorized access to resources
  - ✅ Validates user permissions before operations

  Final Recommendation

  Approve with minor suggestions

  This is an excellent security improvement that addresses critical authentication and authorization gaps. The comprehensive test coverage will prevent security regressions.

  Must-address before merge:
  1. Verify the user.getPermissions → resource.getRolesFor migration doesn't break existing functionality
  2. Add integration test to validate the schema migration works end-to-end

  Nice-to-have improvements:
  1. Extract common test utilities to reduce duplication
  2. Add JSDoc comments explaining the permission model changes

**Review by seaerchin — COMMENTED** [review]
**NOTE: this is a first pass review from claude**

⏺ Code Review: PR #1371 - Security Testing for Routes (Auth + Perms)

  Overview

  This PR adds comprehensive authentication and authorization tests for route handlers across multiple modules (collection, folder, page, resource, user, and me). The changes include 1,744 additions and 291 deletions, focusing on security testing to ensure proper access controls.

  Code Quality & Style

  ✅ Strengths

  - Comprehensive Coverage: Tests cover authentication (401) and authorization (403) scenarios across all major modules
  - Consistent Test Structure: Well-organized test patterns with clear arrange/act/assert structure
  - Security Focus: Addresses critical security concerns by validating permissions at route level
  - Proper Error Handling: Tests verify correct TRPC error codes and messages

  ⚠ Areas for Improvement

  1. Code Duplication in Test Setup
  ```ts
  // Repeated across multiple test files
  it("should throw 401 if not logged in", async () => {
    const result = unauthedCaller.someMethod({...})
    await expect(result).rejects.toThrowError(
      new TRPCError({ code: "UNAUTHORIZED" })
    )
  })
```

  3. Missing Test Utilities
  Tests could benefit from shared utilities for common patterns like permission setup and error assertions.

  4. Schema Changes Need Validation
  - Removal of getPermissionsInputSchema needs verification that it's not breaking existing functionality
  - New schemas like getRolesForSchema should be validated against usage patterns

  Specific Suggestions

  1. Extract Common Test Utilities

```ts
  // tests/integration/helpers/security.ts
  export const expectUnauthorized = (promise: Promise<any>) =>
    expect(promise).rejects.toThrowError(
      new TRPCError({ code: "UNAUTHORIZED" })
    )

  export const expectForbidden = (promise: Promise<any>) =>
    expect(promise).rejects.toThrowError(
      new TRPCError({
        code: "FORBIDDEN",
        message: "You do not have sufficient permissions to perform this action"
      })
    )
```

  2. Improve Query Security

```ts
  // apps/studio/src/server/modules/folder/folder.router.ts:291-294
  const { title } = await db
    .selectFrom("Resource")
    .where("Resource.siteId", "=", siteId) // ✅ Good addition
    .where("Resource.id", "=", resourceId)
    .select("title")
    .executeTakeFirstOrThrow()
```

  3. Schema Migration Validation

  The removal of user.getPermissions route and replacement with resource.getRolesFor needs careful validation:
```ts
  // Before: UserManagementContext.tsx:25
  const { data: roles } = trpc.user.getPermissions.useQuery({ siteId })

  // After: UserManagementContext.tsx:25-28
  const { data: roles } = trpc.resource.getRolesFor.useQuery({
    siteId,
    resourceId: null,
  })
```

  Code Correctness

  - ✅ Proper authentication checks added to all routes
  - ✅ Authorization logic correctly implemented
  - ✅ Database queries properly scoped by siteId
  - ⚠ Schema migration needs validation for breaking changes

  Performance Implications

  - ✅ No significant performance concerns
  - ✅ Database queries are properly indexed fields (siteId, id)
  - ✅ Permission checks are efficient

  Test Coverage

  - ✅ Excellent coverage for auth/authz scenarios
  - ✅ Both positive and negative test cases
  - ✅ Edge cases covered (non-existent resources, wrong permissions)
  - ❌ Missing integration tests for schema migration

  Security Considerations

  - ✅ Major Security Improvement: Adds missing authorization checks
  - ✅ Proper separation of read/write/publish permissions
  - ✅ Prevents unauthorized access to resources
  - ✅ Validates user permissions before operations

  Final Recommendation

  Approve with minor suggestions

  This is an excellent security improvement that addresses critical authentication and authorization gaps. The comprehensive test coverage will prevent security regressions.

  Must-address before merge:
  1. Verify the user.getPermissions → resource.getRolesFor migration doesn't break existing functionality
  2. Add integration test to validate the schema migration works end-to-end

  Nice-to-have improvements:
  1. Extract common test utilities to reduce duplication
  2. Add JSDoc comments explaining the permission model changes

File: apps/studio/src/server/modules/folder/__tests__/folder.router.test.ts [hunk 1]
```diff
@@ -737,6 +737,70 @@ describe("folder.router", async () => {
...
+      await expect(result).rejects.toThrowError(
+        new TRPCError({
+          code: "FORBIDDEN",
+          message:
+            "You do not have sufficient permissions to perform this action",
+        }),
+      )
+    })
+
+    it("should return 200", async () => {
```
[the remark this answers is not in the document]
Comment (adriangohjw) [reply apps/studio/src/server/modules/folder/__tests__/folder.router.test.ts:780]: not quite! i want to limit this PR scope to just make sure we have the most basic forms of auth (login + perms)

File: apps/studio/src/server/modules/permissions/permissions.service.ts [hunk 2]
```diff
@@ -120,18 +120,24 @@ export const validateUserPermissionsForResource = async ({
   }
 }
 
-export const getSitePermissions = async ({
+export const getResourcePermission = async ({
   userId,
   siteId,
-}: Omit<PermissionsProps, "resourceId">) => {
-  return await db
+  resourceId = null,
```
[the remark this answers is not in the document]
Comment (dcshzj) [reply apps/studio/src/server/modules/permissions/permissions.service.ts:126]: @adriangohjw can consider, I think valid point to just make it explicitly optional

File: apps/studio/src/server/modules/site/site.router.ts [hunk 3]
```diff
@@ -181,163 +175,166 @@ export const siteRouter = router({
...
+          }
+
+          const newNavbar = await tx
+            .updateTable("Navbar")
+            .set({
+              content: jsonb(
+                safeJsonParse(
+                  navbar,
+                ) as IsomerSiteWideComponentsProps["navbar"],
+              ),
```
[the remark this answers is not in the document]
Comment (dcshzj) [reply apps/studio/src/server/modules/site/site.router.ts:263]: Nonsense, ignore

File: apps/studio/src/server/modules/collection/collection.router.ts [hunk 4]
```diff
@@ -259,6 +257,7 @@ export const collectionRouter = router({
       const baseQuery = db
         .selectFrom("Resource")
         .where("Resource.id", "=", String(linkId))
```
[the remark this answers is not in the document]
Comment (dcshzj) [reply apps/studio/src/server/modules/collection/collection.router.ts:259]: Nonsense, ignore

**Review by dcshzj — APPROVED** [review]
